HDDS-9296. Add Swagger UI for Recon REST APIs

devabhishekpal opened the pull request:
## What changes were proposed in this pull request?
Currently the documentation provided for Recon APIs do not have a Swagger UI for the same.
This PR introduces the proposed Swagger UI, providing an interactive documentation for the available APIs and the responses it provides.

* The Swagger UI is a form of documentation that would allow the users to interact with the documentation that is provided for the Recon APIs.
* This would allow people to quickly interact with these endpoints and understand the type of payload to be given, the expected response and it's structure
* It is a supplemental documentation over the existing Recon API document that is present.
### Quick overview for the YAML file format:
* The YAML file to store the definitions for the APIs has the following main things to note:

| Section Name | Use |
|---------------|-------|
| tags | Stores the names by which we can group APIs/data, for example Containers and Keys, Namespace etc. |
| paths | Stores the actual API definition data and information like short descriptions, accepted parameters for the request, response data. The responses have been defined in the components section, and `$ref` is used to provide a reference to allow readability and organization in the code |
| components | Stores the actual format for the data response |

#### `path` section has the following main information
| Sub-section | Use |
|-------------|------|
| paths / `get, post, put...` | Under path, the get word defines what type of request the API will accept. In this case we have given get for the APIs but we can also do post, put, delete etc. |
| paths / `<request-type>` / tags | Specifies which group the API should be put under |
| paths / `<request-type>` / parameters | This section would be used to define the various parameters that would be accepted by the API for the request |
| paths / `<request-type>` / responses | This section contains the information about the response that would be given. Notice we write the information under a '200' section which means when the API returns a 200 status the further definition for the response would be used. We can further expand this to hold response types for other statuses as well |
| paths / ... / responses / content | Stores the actual response type. Here we have defined the response to be of type ` application/json`, but we can have other types as well like `application/xml` |
| paths / ... / responses / content / ... / schema | This holds a reference to the actual structure of the response which we have defined in the `components` section |

#### `components` section has the following main information
| Sub-section | Use |
|-------------|------|
| component / schema | This will mainly be used for data that might have a fixed format, similar to a documentation containing syntax information or information on object. Note: This also allows users to see the response format without looking at the actual response as it will define the possible fields and their types in a compact manner |
| component / responses | This will be used to store responses directly. The data in this section will not be rendered in the Schemas dropdown present in the UI |

#### Overview on the data-types for the responses
The YAML response format is as follows:
* type: <type of data >
* example: <a sample value to be displayed>

Now the data types supported are all the general primitive data-types we use:
* string
* number / integer -> Both of them hold a numeric value, however number supports larger values than the typical integer range and also floating point values
* boolean
* array
* object

For a *data-type of array* we need to include the following information as well:
* items: This will define the possible items that would be stored in the array and stores the format for those items.
For a *data-type of object* we need to include the following information as well:
* properties: This will define the keys / properties that are associated with the object.
    * Ex:
```
properties:
    a: ....
    b: .... 
```
would correspond to an object like:
```
{
    "a": .... ,
    "b": ....
}
```
We can further chain these to create complex data types like an array of objects, or an object key of type array which can be seen in the yaml file

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9296

## How was this patch tested?
This patch was tested manually. Following are the related screenshots and recording for the documentation page.

<img width="1728" alt="HDDS-9296-ss_1" src="https://github.com/apache/ozone/assets/43001336/4102a88a-cf27-47ae-9682-b8dd08e9e600">
<img width="1728" alt="HDDS-9296-ss_2" src="https://github.com/apache/ozone/assets/43001336/e3ad8dd5-c64b-489a-96b4-eaebb866307e">
<img width="1728" alt="HDDS-9296-ss_3" src="https://github.com/apache/ozone/assets/43001336/acf6a998-ab5a-4a1d-8f7a-1153945b5ce7">

https://github.com/apache/ozone/assets/43001336/a59df031-23da-4e8c-a245-f5dd36265d1c

adoroszlai commented:
@devmadhuu @ArafatKhan2198 please review

devabhishekpal commented:
Hello @ivandika3 could you also please take a look at this?

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+      tags:
+        - Containers and Keys
+      summary: Get all Container Metadata information
+      operationId: getContainerInfo
```
How this operationId gets generated ?

devabhishekpal replied:
So operationId will not be displayed in the UI, it is a unique identifier internally for Swagger, so that in future if we want to reference some api we can use that.


----
From [swagger docs](https://swagger.io/docs/specification/paths-and-operations/):
Some common use cases for operationId are:

* Some code generators use this value to name the corresponding methods in code.
* [Links](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#linkObject) can refer to the linked operations by operationId.

devmadhuu replied:
> So operationId will not be displayed in the UI, it is a unique identifier internally for Swagger, so that in future if we want to reference some api we can use that.
> 
> From [swagger docs](https://swagger.io/docs/specification/paths-and-operations/): Some common use cases for operationId are:
> 
> * Some code generators use this value to name the corresponding methods in code.
> * [Links](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#linkObject) can refer to the linked operations by operationId.

Can we change it to reflect method name ?

devabhishekpal replied:
Sure, we can do that

devabhishekpal replied:
Addressed in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+      summary: Get the Key metadata for all keys present in the given container ID
+      operationId: getContainerKeyMetadata
+      parameters:
+        - name: containerId
```
Pls check this param name. Seems not correct.

devabhishekpal replied:
So in the API from the docs, we have the parameter as `:id` which is acting as a placeholder to denote the id of the container. We cannot however refer to this in the URL like that. But we can replace containerId with just id to be more in line with the documentation.
It was given as containerId for better readability.

devmadhuu replied:
> So in the API from the docs, we have the parameter as `:id` which is acting as a placeholder to denote the id of the container. We cannot however refer to this in the URL like that. But we can replace containerId with just id to be more in line with the documentation. It was given as containerId for better readability.

If possible, make it inline with code, as Swagger yaml also works as for doc purposes as an open standard. And description of param may reflect the meaning of it.

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ContainerMetadata'
+  /containers/{containerId}/keys:
```
Correct the `containerId` param here

devabhishekpal replied:
So here `containerId` is a placeholder for `id` in the documentation:
`/api/v1/containers/:id/keys`

Swagger does not support naming of paths like that.
From [swagger docs](https://swagger.io/docs/specification/describing-parameters/#path-parameters):

```Path parameters are variable parts of a URL path. They are typically used to point to a specific resource within a collection, such as a user identified by ID. A URL can have several path parameters, each denoted with curly braces { }.```

----
What this basically means is, when we are interacting with the swagger UI, if we click `Try It Out`, we can provide parameters to the UI, check in the attached screenshot containerId is provided as `111`.
Now in the sample cURL call provided (red highlight), you will notice the url as:
`'http://localhost:1313/api/v1/containers/111/keys.....`
<img width="1728" alt="Screenshot 2023-11-03 at 21 13 50" src="https://github.com/apache/ozone/assets/43001336/e5b678e3-73ed-4538-bf4d-7133b4c0004e">

Here the 111 is defined to be placed as a path parameter, and is replaced in the `{containerId}` part of the URL.
The second type is the query parameter, which appears as `?prevKey=%2Fvol1%2Fbucket1%2Fkey1&limit=1000` in the URL.

devmadhuu replied:
> So here `containerId` is a placeholder for `id` in the documentation: `/api/v1/containers/:id/keys`
> 
> Swagger does not support naming of paths like that. From [swagger docs](https://swagger.io/docs/specification/describing-parameters/#path-parameters):
> 
> `Path parameters are variable parts of a URL path. They are typically used to point to a specific resource within a collection, such as a user identified by ID. A URL can have several path parameters, each denoted with curly braces { }.`
> 
> What this basically means is, when we are interacting with the swagger UI, if we click `Try It Out`, we can provide parameters to the UI, check in the attached screenshot containerId is provided as `111`. Now in the sample cURL call provided (red highlight), you will notice the url as: `'http://localhost:1313/api/v1/containers/111/keys.....` <img alt="Screenshot 2023-11-03 at 21 13 50" width="1728" src="https://user-images.githubusercontent.com/43001336/280336503-e5b678e3-73ed-4538-bf4d-7133b4c0004e.png">
> 
> Here the 111 is defined to be placed as a path parameter, and is replaced in the `{containerId}` part of the URL. The second type is the query parameter, which appears as `?prevKey=%2Fvol1%2Fbucket1%2Fkey1&limit=1000` in the URL.

Not necessarily the code format, just `id` inline with code makes sense along with description of param. If possible, then inline with code makes more sense.

devabhishekpal replied:
Reverted to {id} in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+  /containers/missing:
+    get:
+      tags:
+        - Containers and Keys
```
not sure if keys should be included as tag here.

devabhishekpal replied:
So this tag was added as the current recon doc has this specified under Containers and Keys section.
Ref: https://ozone.apache.org/docs/edge/interface/reconapi.html
We can separate it out as Containers only and Keys only

devmadhuu replied:
> So this tag was added as the current recon doc has this specified under Containers and Keys section. Ref: https://ozone.apache.org/docs/edge/interface/reconapi.html We can separate it out as Containers only and Keys only

Yeah, I meant if API relates to any operation for keys and containers, then having both makes sense, else we can separate out and keep combined one as well.

devabhishekpal replied:
Sure, addressed in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/MissingContainerMetadata'
+  /containers/{containerId}/replicaHistory:
```
Correct the `containerId` param here

devabhishekpal replied:
Please check the reply at https://github.com/apache/ozone/pull/5509#discussion_r1381907219 for the same

devmadhuu replied:
> Please check the reply at [#5509 (comment)](https://github.com/apache/ozone/pull/5509#discussion_r1381907219) for the same

Pls check above comments and similar for below.

devabhishekpal replied:
Addressed in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+  /containers/{containerId}/replicaHistory:
+    get:
+      tags:
+        - Containers and Keys
```
keys should not be included as tag here.

devabhishekpal replied:
Please check https://github.com/apache/ozone/pull/5509#discussion_r1381907931 and let me know if that change would be okay

devabhishekpal replied:
Addressed in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+      summary: Get all the Replica history for a given container id
+      operationId: getReplicaHistory
+      parameters:
+        - name: containerId
```
Pls check this param name. Seems not correct.

devabhishekpal replied:
This is the same as https://github.com/apache/ozone/pull/5509#discussion_r1381910275

devabhishekpal replied:
Reverted to {id} in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+  /containers/unhealthy:
+    get:
+      tags:
+        - Containers and Keys
```
Same comment here as above.

devabhishekpal replied:
Addressed in the latest commit

devmadhuu commented on hadoop-hdds/docs/themes/ozonedoc/static/swagger-resources/recon-api.yaml:
```diff
@@ -0,0 +1,1584 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/UnhealthyContainerMetadata'
+  /containers/unhealthy/{containerState}:
```
check the param `containerState`, should be `state`

devabhishekpal replied:
Sure, this is the same as https://github.com/apache/ozone/pull/5509#discussion_r1381910275 but we can change it

devmadhuu replied:
> Sure, this is the same as [#5509 (comment)](https://github.com/apache/ozone/pull/5509#discussion_r1381910275) but we can change it

Inline with code looks uniform as Swagger yaml being open standard , so not to just think about link, I feel we should have it in line with code along with param description.

devabhishekpal replied:
Addressed in the latest commit, reverted to {state}

devmadhuu reviewed (commented):
@devabhishekpal  thanks for working on this patch. Found anamolies in recon-swagger.yaml w.r.t param names correctness. Pls check once and one round of self review with actual APIs

ArafatKhan2198 reviewed (commented):
Thanks for working on this @devabhishekpal 
I am not able to access the endpoints from the link it is throwing an error.
I followed the steps you showed in the posted video am I missing something here?
<img width="871" alt="image" src="https://github.com/apache/ozone/assets/98023601/1d2e00ba-a7af-462e-9f27-0df8a8414ab1">

devabhishekpal commented:
Thanks a lot for reviewing these changes @devmadhuu .
In the latest commit I did the changes as requested.
Updated screenshots:
<img width="1728" alt="Screenshot 2023-11-03 at 23 50 31" src="https://github.com/apache/ozone/assets/43001336/3b55f84d-88a0-4187-b13f-1616d0f1af31">
<img width="1728" alt="Screenshot 2023-11-03 at 23 51 25" src="https://github.com/apache/ozone/assets/43001336/622fecbe-9b10-4341-88fc-6af896a8ede3">
<img width="1728" alt="Screenshot 2023-11-03 at 23 58 08" src="https://github.com/apache/ozone/assets/43001336/bdfd6440-e4d9-4095-8030-2dbd39768cf8">

I corrected the `operationId` to reflect the method names as declared in the methods from the [API folder](https://github.com/apache/ozone/tree/master/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api).
However for the `/containers/unhealthy/{state}` the operation ID is different from the method name as both the endpoints i.e.
* /containers/unhealthy
* /containers/unhealthy/{state}
have the same method name, causing Swagger to trigger both dropdowns.

I put the **operationId for /containers/unhealthy** as `getUnhealthyContainers` which is it's method name.
And for  **/containers/unhealthy/{state}**, I put it as `getUnhealthyContainersWithState` to make it unique.

devmadhuu reviewed (approved):
Thanks @devabhishekpal for addressing the review comments. PR LGTM +1 . Pls also get reviewed by someone.

ArafatKhan2198 commented on hadoop-hdds/docs/content/interface/SwaggerReconApi.md:
```diff
@@ -0,0 +1,25 @@
+  limitations under the License.
+-->
+
+{{< swagger-ui url="/swagger-resources/recon-api.yaml" >}}
```
We have set this path as a hardcoded value, this can change from env to env right?
For example in my docker setup we have this file placed at the following path :- 
```
/docs/swagger-resources/recon-api.yaml 
```
We need to set it dynamically according to the prod env.

devabhishekpal replied:
Hello @ArafatKhan2198 , thanks for asking the question. While it is true that we would need to handle this kind of scenario based on the prod env, I think for docs we do not build it as a part of the overall Ozone parcel, but rather in the ozone-site repo via GitHub actions.

After testing it out, it seems for the hosted site, the relative path would be:
`/docs/egde/swagger-resources/recon-api.yaml`

<img width="1728" alt="Screenshot 2023-11-06 at 20 30 28" src="https://github.com/apache/ozone/assets/43001336/acdf5997-f19c-4832-a43b-df9c393cd14c">

How I tested it is:
* Ran the following in the Ozone main repo:
```
mvn --no-transfer-progress -pl :hdds-docs clean package
```
* Built the master branch for `ozone-site` repo using `hugo`
* In the public folder that was generated, I created a folder named **edge**
* Copied the files from `hadoop-hdds/docs/target/classes/docs` in the Ozone repo to the above public folder.
* Hosted the site locally and tested the location manually.

@adoroszlai @errose28 This would be the correct way that the ozone-site gets generated right? So this location should be good.

adoroszlai commented:
`docs/edge` on the website is just a convenience for developers to have a snapshot of the current docs.  Docs are also built during release, and are part of the binary tarball users can [download from the website](https://ozone.apache.org/downloads/).  They can extract the binaries to any local dir.  Is it possible to make the Swagger UI work for them?

devabhishekpal commented:
Hello, thanks for the review @adoroszlai @ArafatKhan2198. In the latest commit I changed it to a relative path, which should allow the yaml file to be detected irrespective of how the files are being stored.
Verified this patch for the following:
* By running the same steps as in https://github.com/apache/ozone/pull/5509#discussion_r1383483505
<img width="1728" alt="Screenshot 2023-11-06 at 21 38 32" src="https://github.com/apache/ozone/assets/43001336/a536bdb9-5136-430a-a98c-e5fe62d00fa7">
* By directly running `hugo serve` on the docs folder
<img width="1728" alt="Screenshot 2023-11-06 at 21 40 53" src="https://github.com/apache/ozone/assets/43001336/9c319c77-cc9c-4efb-a97c-46a50fcf9d10">
* Via building the whole project, running it using `docker-compose` and then visiting the UI
<img width="1728" alt="Screenshot 2023-11-06 at 22 17 00" src="https://github.com/apache/ozone/assets/43001336/2c4ec1e2-e24c-4e87-a1d6-60686325a807">

ArafatKhan2198 commented:
> Hello, thanks for the review @adoroszlai @ArafatKhan2198. In the latest commit I changed it to a relative path, which should allow the yaml file to be detected irrespective of how the files are being stored. Verified this patch for the following:
> 
> * By running the same steps as in [HDDS-9296. Add Swagger UI for Recon REST APIs #5509 (comment)](https://github.com/apache/ozone/pull/5509#discussion_r1383483505)
> 
> <img alt="Screenshot 2023-11-06 at 21 38 32" width="1728" src="https://user-images.githubusercontent.com/43001336/280773472-a536bdb9-5136-430a-a98c-e5fe62d00fa7.png">
> * By directly running `hugo serve` on the docs folder
> 
> <img alt="Screenshot 2023-11-06 at 21 40 53" width="1728" src="https://user-images.githubusercontent.com/43001336/280773929-9c319c77-cc9c-4efb-a97c-46a50fcf9d10.png">
> * Via building the whole project, running it using `docker-compose` and then visiting the UI
> 
> <img alt="Screenshot 2023-11-06 at 22 17 00" width="1728" src="https://user-images.githubusercontent.com/43001336/280783775-2c4ec1e2-e24c-4e87-a1d6-60686325a807.png">

Thanks for the changes @devabhishekpal Ill take a look!

errose28 reviewed (commented):
Thanks for adding this @devabhishekpal. I previewed the changes locally by running `hugo serve` from the *hadoop-hdds/docs* directory and it rednered fine for me. Should we disable the "try it out" button (I think that's possible [here](https://github.com/swagger-api/swagger-ui/blob/84de63c5bcc57c4fbdad6d41b270faf1ba911456/docs/usage/configuration.md?plain=1#L239)) since the requests are not going to know where the running recon instance is or have the ability to authenticate?

devabhishekpal commented:
Hello @errose28, so while it is true that "Try It Out" button will be giving error since there is no recon instance, the main reason for the button is to provide the interactivity.
Without it the documentation would be equivalent to the normal documentation we have, only with a different UI.
On clicking **Try It Out**, the end user can interact with the various params, and even get a URL for the API call with the params filled in.
<img width="1728" alt="Screenshot 2023-11-09 at 12 54 21" src="https://github.com/apache/ozone/assets/43001336/70e0f93d-53e4-4f7a-9de1-f296d19808d6">

That is also one of the reasons to add a Swagger UI.

ArafatKhan2198 commented:
@devabhishekpal thanks for making the changes! Quick question?
Is the schema generated automatically by using the example response, or do you need to write it down manually?

<img width="531" alt="image" src="https://github.com/apache/ozone/assets/98023601/c7306341-d9bd-4d05-8c37-bae9cc648c17">

devabhishekpal commented:
@ArafatKhan2198 this schema is generated automatically from the example and schema structure declared.

So if you check the yaml file we have the schema declared as:

```
someDataObject:
    type: object
    properties:
        a:
         type: integer
         example: 3
```

So from here it interpolates the schema

ArafatKhan2198 reviewed (changes requested):
### 1. Containers : 
1. The default value for `/containers/mismatch` also mention in description. 
    1. prevKey = 0 
    2. missingIn = SCM
 2. The default value for `/containers/mismatch/deleted` also mention in description. 
    1. prevKey = 0

ArafatKhan2198 reviewed (changes requested):
### 2. Keys :
1. The port on which Recon is running is not configured correctly when we test the API feature using Swagger's "try out" feature. In the local Docker setup, it defaults to using port 9888, but it can be configured according to the user's preferences. We cannot set a fixed value for it, it needs to be listen to the port the user has set. This is not just for Open keys but for all other API's in general.
<img width="1603" alt="image" src="https://github.com/apache/ozone/assets/98023601/0dd8e081-0434-402f-b529-447743deefbc">

2. For Open Keys we have two more query parameter ➖`includeFso` and `includeNonFso` which are boolean variables which by default both are set to `false`
<img width="1651" alt="image" src="https://github.com/apache/ozone/assets/98023601/f0de1358-ff75-459f-bf0d-78df113d7c3e">

ArafatKhan2198 reviewed (changes requested):
### 3. Containers & Keys ➖

1. For the endpoint `/containers/{id}/keys` please change the `prevKey` to empty string for default value.
<img width="602" alt="image" src="https://github.com/apache/ozone/assets/98023601/7b9a5224-1dac-4c7b-8e50-ceccb85b2fa3">

ArafatKhan2198 reviewed (changes requested):
### 4. Namespace Metadata ➖ 
1. Can you change the description for the `files` query parameter to “_This boolean has a default value of false. When set to true, it calculates the disk usage for keys within the specified path and also provides a list of their corresponding sub-paths_.”

<img width="1608" alt="image" src="https://github.com/apache/ozone/assets/98023601/ec02b2ac-e6ae-4ddc-940d-fa67b6523693">

ArafatKhan2198 reviewed (changes requested):
### 5. Utilization ➖
1. Change the description for `fileSize` query parameter for `/utilization/fileCount` to the following :- _" Filters the results based on the given file size. The smallest file size being tracked for count is 1 KB i.e. 1024 bytes."_
<img width="465" alt="image" src="https://github.com/apache/ozone/assets/98023601/fd06e627-e04a-424d-8a45-213d5d79959c">

2. We have also included another API in the utilization endpoints, specifically for container counts. However, it has not been added to the official Apache documentation at this time. You can take the following information and add it to the response doc :- 

```
Endpoint :- /utilization/containerCount?containerSize=512000000

Parameters :- 
Name :- containerSize 
Description :- Filters the results based on the given container size. The smallest container size being tracked for count is 512 MB i.e. 512000000 bytes.
Example :- 512000000

Response :- 
[
  {
    "containerSize": 2147483648,
    "count": 9
  },
  {
    "containerSize": 1073741824,
    "count": 3
  }
]
```

ArafatKhan2198 commented:
Thanks for the changes @devabhishekpal I have left comments.

devabhishekpal commented:
Thanks @ArafatKhan2198 for the review.
I have made the changes.
The only thing which isn't quite feasible is from https://github.com/apache/ozone/pull/5509#pullrequestreview-1725354077 regarding setting of the port value.

Swagger will determine the port from the localhost value. While we can do that via JS, maybe add an option for the users to specify the end-port, most users might not use it, since we are returning 404 either way because of no recon instance, and this is just an interactive documentation - not a replacement for API calls.

But do let me know if you want that change, will look into it.

The other changes have been addressed in the latest commit.
Please do take a look once, specially in the `/utilization/containerCount` API, and let me know if the description is okay, and whether containerSize parameter is mandatory or not.

ArafatKhan2198 commented:
Thank you, @devabhishekpal, for sticking with this and implementing the necessary changes. Just a few minor comments, and we should be ready to proceed with this.

1. For the `/utilization/containerCount` endpoint the parameter `containerSize` is not a required field and is just optional. 
<img width="584" alt="image" src="https://github.com/apache/ozone/assets/98023601/389af691-c99f-48d7-9c7c-7cffb792d0da">

2. Could you please add this message next to the `containerSize` parameter, similar to how it is done for the fileSize? Its better we maintain consistency throughout ➖
<img width="1631" alt="image" src="https://github.com/apache/ozone/assets/98023601/aab5e4fe-b1f5-4342-8ef8-e7afcd717bbb">

ArafatKhan2198 reviewed (approved):
Thanks for taking this up @devabhishekpal appreciate the effort :)
LGTM +1

errose28 reviewed (approved):
Thanks for the effort assembling this @devabhishekpal and the thorough review @ArafatKhan2198. LGTM as well.

dombizita reviewed (approved):
Thank you for your work on this @devabhishekpal, it looks good to me as well!

dombizita commented:
Thanks for working on this @devabhishekpal! Thanks for the review @devmadhuu, @ArafatKhan2198, @adoroszlai and @errose28!

smengcl commented:
Thanks @devabhishekpal for picking this one up!
